File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/javascript/find_first_duplicate.js
function findFirstDuplicate(arr) {
  const uniques = new Set();

  for (const value of arr) {
    if (uniques.has(value)) {
      return value;
    }
    uniques.add(value);
  }

  return -1;
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));
}

module.exports = findFirstDuplicate;


// sets contain only unique values; check value in array and add to set if it doesnt exist in set, otherwise return duplicate. return -1 when no duplicate to return 

File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/ruby/find_first_duplicate.rb
require 'set'

def find_first_duplicate(arr)
  uniques = Set.new

  arr.each do |value|
    return value if uniques.include?(value)

    uniques.add(value)
  end

  -1
end


if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  # Don't forget to add your own!
end

# Please add your pseudocode to this file

# sets contain only unique values; check value in array and add to set if it doesnt exist in set, otherwise return duplicate. return -1 when no duplicate to return